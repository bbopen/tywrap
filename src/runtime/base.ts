/**
 * Base runtime bridge
 */

import type { RuntimeExecution } from '../types/index.js';

export abstract class RuntimeBridge implements RuntimeExecution {
  abstract call<T = unknown>(
    module: string,
    functionName: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T>;

  abstract instantiate<T = unknown>(
    module: string,
    className: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T>;

  abstract dispose(): Promise<void>;
}
