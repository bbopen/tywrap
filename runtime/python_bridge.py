#!/usr/bin/env python3
import sys
import json
import importlib
import os
import traceback

instances = {}

FALLBACK_JSON = os.environ.get('TYWRAP_CODEC_FALLBACK', '').lower() == 'json'


def serialize(obj):
    try:
        import numpy as np  # noqa: F401
        if hasattr(obj, 'tolist') and hasattr(obj, 'shape'):
            try:
                import pyarrow as pa  # type: ignore
                import base64
                # numpy array -> Arrow tensor (fallback to tolist when not supported)
                try:
                    arr = pa.array(obj)
                    sink = pa.BufferOutputStream()
                    with pa.ipc.new_stream(sink, arr.type) as writer:
                        writer.write(arr)
                    buf = sink.getvalue()
                    b64 = base64.b64encode(buf.to_pybytes()).decode('ascii')
                    return { '__tywrap__': 'ndarray', 'encoding': 'arrow', 'b64': b64, 'shape': getattr(obj, 'shape', None) }
                except Exception:
                    if not FALLBACK_JSON:
                        raise
            except Exception:
                if not FALLBACK_JSON:
                    raise
            if FALLBACK_JSON:
                return { '__tywrap__': 'ndarray', 'encoding': 'json', 'data': obj.tolist(), 'shape': getattr(obj, 'shape', None) }
            raise RuntimeError('Arrow encoding unavailable for ndarray; install pyarrow or set TYWRAP_CODEC_FALLBACK=json to enable JSON fallback')
    except Exception:
        pass
    try:
        import pandas as pd  # noqa: F401
        # Pandas DataFrame
        if hasattr(obj, 'to_dict') and getattr(type(obj), '__name__', '') == 'DataFrame':
            try:
                import pyarrow as pa  # type: ignore
                import pyarrow.feather as feather  # type: ignore
                import base64
                try:
                    table = pa.Table.from_pandas(obj)  # type: ignore
                    sink = pa.BufferOutputStream()
                    feather.write_feather(table, sink)
                    buf = sink.getvalue()
                    b64 = base64.b64encode(buf.to_pybytes()).decode('ascii')
                    return { '__tywrap__': 'dataframe', 'encoding': 'arrow', 'b64': b64 }
                except Exception:
                    if not FALLBACK_JSON:
                        raise
            except Exception:
                if not FALLBACK_JSON:
                    raise
            if FALLBACK_JSON:
                return { '__tywrap__': 'dataframe', 'encoding': 'json', 'data': obj.to_dict(orient='records') }
            raise RuntimeError('Arrow encoding unavailable for pandas.DataFrame; install pyarrow or set TYWRAP_CODEC_FALLBACK=json to enable JSON fallback')
        # Pandas Series
        if getattr(type(obj), '__name__', '') == 'Series':
            try:
                import pyarrow as pa  # type: ignore
                import base64
                try:
                    arr = pa.Array.from_pandas(obj)  # type: ignore
                    sink = pa.BufferOutputStream()
                    with pa.ipc.new_stream(sink, arr.type) as writer:
                        writer.write(arr)
                    buf = sink.getvalue()
                    b64 = base64.b64encode(buf.to_pybytes()).decode('ascii')
                    return { '__tywrap__': 'series', 'encoding': 'arrow', 'b64': b64, 'name': getattr(obj, 'name', None) }
                except Exception:
                    if not FALLBACK_JSON:
                        raise
            except Exception:
                if not FALLBACK_JSON:
                    raise
            try:
                data = obj.to_list()  # type: ignore
            except Exception:
                data = obj.to_dict()  # type: ignore
            if FALLBACK_JSON:
                return { '__tywrap__': 'series', 'encoding': 'json', 'data': data, 'name': getattr(obj, 'name', None) }
            raise RuntimeError('Arrow encoding unavailable for pandas.Series; install pyarrow or set TYWRAP_CODEC_FALLBACK=json to enable JSON fallback')
    except Exception:
        pass
    return obj

def handle_call(params):
    module_name = params.get('module')
    function_name = params.get('functionName')
    args = params.get('args') or []
    kwargs = params.get('kwargs') or {}
    mod = importlib.import_module(module_name)
    func = getattr(mod, function_name)
    res = func(*args, **kwargs)
    return serialize(res)

def handle_instantiate(params):
    module_name = params.get('module')
    class_name = params.get('className')
    args = params.get('args') or []
    kwargs = params.get('kwargs') or {}
    mod = importlib.import_module(module_name)
    cls = getattr(mod, class_name)
    obj = cls(*args, **kwargs)
    handle_id = str(id(obj))
    instances[handle_id] = obj
    return handle_id

def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
            mid = msg.get('id')
            method = msg.get('method')
            params = msg.get('params') or {}
            try:
                if method == 'call':
                    result = handle_call(params)
                elif method == 'instantiate':
                    result = handle_instantiate(params)
                else:
                    raise ValueError('Unknown method')
                out = { 'id': mid, 'result': result }
            except Exception as e:
                out = {
                    'id': mid,
                    'error': { 'type': type(e).__name__, 'message': str(e), 'traceback': traceback.format_exc() }
                }
        except Exception as e:
            out = { 'id': -1, 'error': { 'type': type(e).__name__, 'message': str(e) } }
        sys.stdout.write(json.dumps(out) + '\n')
        sys.stdout.flush()

if __name__ == '__main__':
    main()


