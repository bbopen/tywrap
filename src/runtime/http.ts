/**
 * HTTP runtime bridge
 */

import { RuntimeBridge } from './base.js';

export interface HttpBridgeOptions {
  baseURL: string;
  headers?: Record<string, string>;
  timeoutMs?: number;
}

interface HttpCallPayload {
  module: string;
  functionName: string;
  args: unknown[];
  kwargs?: Record<string, unknown>;
}

interface HttpInstantiatePayload {
  module: string;
  className: string;
  args: unknown[];
  kwargs?: Record<string, unknown>;
}

export class HttpBridge extends RuntimeBridge {
  private readonly baseURL: string;
  private readonly headers: Record<string, string>;
  private readonly timeoutMs: number;

  constructor(options: HttpBridgeOptions = { baseURL: 'http://localhost:8000' }) {
    super();
    this.baseURL = options.baseURL.replace(/\/$/, '');
    this.headers = { 'content-type': 'application/json', ...(options.headers ?? {}) };
    this.timeoutMs = options.timeoutMs ?? 30000;
  }

  async call<T = unknown>(
    module: string,
    functionName: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T> {
    const payload: HttpCallPayload = { module, functionName, args, kwargs };
    const res = await this.post(`${this.baseURL}/call`, payload);
    return res as T;
  }

  async instantiate<T = unknown>(
    module: string,
    className: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T> {
    const payload: HttpInstantiatePayload = { module, className, args, kwargs };
    const res = await this.post(`${this.baseURL}/instantiate`, payload);
    return res as T;
  }

  async dispose(): Promise<void> {
    // stateless
  }

  private async post(url: string, body: unknown): Promise<unknown> {
    const controller = typeof AbortController !== 'undefined' ? new AbortController() : undefined;
    const timer = controller ? setTimeout(() => controller.abort(), this.timeoutMs) : undefined;
    try {
      const resp = await fetch(url, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(body),
        signal: controller?.signal,
      });
      if (!resp.ok) {
        const text = await safeText(resp);
        throw new Error(`HTTP ${resp.status}: ${text || resp.statusText}`);
      }
      const ct = resp.headers.get('content-type') ?? '';
      if (ct.includes('application/json')) {
        return (await resp.json()) as unknown;
      }
      const text = await resp.text();
      try {
        return JSON.parse(text) as unknown;
      } catch {
        return text as unknown;
      }
    } finally {
      if (timer) {
        clearTimeout(timer);
      }
    }
  }
}

async function safeText(resp: Response): Promise<string> {
  try {
    return await resp.text();
  } catch {
    return '';
  }
}
