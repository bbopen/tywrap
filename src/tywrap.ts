/**
 * Main tywrap API entry point
 */

import { CodeGenerator } from './core/generator.js';
import { TypeMapper } from './core/mapper.js';
import type {
  TywrapOptions,
  PythonFunction,
  PythonModule as TSPythonModule,
  PythonClass,
  Parameter,
  PythonType,
} from './types/index.js';
import { fsUtils, pathUtils, processUtils, hashUtils } from './utils/runtime.js';

// Collect unknown typing constructs encountered during annotation parsing (per-generate run)
let unknownTypeNamesCollector: Map<string, number> = new Map();
function recordUnknown(name: string): void {
  const prev = unknownTypeNamesCollector.get(name) ?? 0;
  unknownTypeNamesCollector.set(name, prev + 1);
}

/**
 * Main tywrap function
 */
export interface TywrapInstance {
  mapper: TypeMapper;
  generator: CodeGenerator;
  options: Partial<TywrapOptions>;
}

export async function tywrap(options: Partial<TywrapOptions> = {}): Promise<TywrapInstance> {
  const mapper = new TypeMapper();
  const generator = new CodeGenerator();

  return {
    mapper,
    generator,
    options,
  };
}

export default tywrap;

/**
 * Append minimal tsd tests for a generated module (optional dev aid)
 */
export async function emitTypeTestsForModule(
  moduleName: string,
  outDir = 'test-d/generated'
): Promise<void> {
  const { writeFile, mkdir } = await import('fs/promises');
  const { join } = await import('path');
  try {
    await mkdir(outDir, { recursive: true });
  } catch {}
  const filePath = join(outDir, `${moduleName}.test-d.ts`);
  const content = `import { expectType } from 'tsd';
import * as mod from '../../generated/${moduleName}.generated.ts';

// Opportunistic: ensure module namespace is an object
expectType<Record<string, unknown>>(mod);
`;
  await writeFile(filePath, content, 'utf-8');
}

/**
 * Generate TypeScript wrappers for configured Python modules.
 * Minimal MVP implementation: resolves module file, analyzes, generates TS, writes to output.dir
 */
export async function generate(
  options: Partial<TywrapOptions>
): Promise<{ written: string[]; warnings: string[] }> {
  const instance = await tywrap(options);
  const written: string[] = [];
  const warnings: string[] = [];
  const outputDir = options.output?.dir ?? './generated';
  const caching = options.performance?.caching ?? false;
  const cacheDir = '.tywrap/cache';

  // Ensure directory exists (Node-only best-effort)
  try {
    const modFs = await import('fs/promises');
    await modFs.mkdir(outputDir, { recursive: true });
  } catch {
    // ignore in non-node or if already exists
  }

  const modules = options.pythonModules ?? {};
  for (const entry of Object.entries(modules)) {
    const moduleKey = entry[0];
    // reset collector for each module
    unknownTypeNamesCollector = new Map();
    // Prefer Python IR extractor over TS analyzer with optional cache
    const cacheKey = await computeCacheKey(moduleKey, options);
    let ir: unknown | null = null;
    if (caching && fsUtils.isAvailable()) {
      try {
        const modFs = await import('fs/promises');
        await modFs.mkdir(cacheDir, { recursive: true });
      } catch {}
      try {
        const cached = await fsUtils.readFile(pathUtils.join(cacheDir, cacheKey));
        ir = JSON.parse(cached);
      } catch {
        ir = null;
      }
    }
    if (!ir) {
      const pythonPath = options.runtime?.node?.pythonPath ?? 'python3';
      ir = await fetchPythonIr(moduleKey, pythonPath);
      // Basic warning when stderr occurred is handled inside fetch; here we simply note miss
      if (ir && caching && fsUtils.isAvailable()) {
        try {
          await fsUtils.writeFile(pathUtils.join(cacheDir, cacheKey), JSON.stringify(ir));
        } catch {}
      }
    }
    if (!ir) {
      warnings.push(`No IR produced for module ${moduleKey}`);
      continue;
    }

    const moduleModel = transformIrToTsModel(ir);

    // Generate module code
    const annotatedJSDoc = Boolean(options.output?.annotatedJSDoc);
    const gen = instance.generator.generateModuleDefinition(moduleModel, annotatedJSDoc);

    // Write file
    const baseName = moduleModel.name || 'module';
    const outPath = pathUtils.join(outputDir, `${baseName}.generated.ts`);
    await fsUtils.writeFile(outPath, gen.typescript);
    written.push(outPath);

    // Optional .d.ts emission (header-only declarations mirroring exports)
    if (options.output?.declaration) {
      const dtsPath = pathUtils.join(outputDir, `${baseName}.generated.d.ts`);
      const dts = renderDts(gen.typescript);
      await fsUtils.writeFile(dtsPath, dts);
      written.push(dtsPath);
    }

    // Optional source map emission (placeholder mapping for now)
    if (options.output?.sourceMap) {
      const mapPath = pathUtils.join(outputDir, `${baseName}.generated.ts.map`);
      const map = renderSourceMapPlaceholder(moduleModel.name);
      await fsUtils.writeFile(mapPath, map);
      written.push(mapPath);
    }

    // Emit warning summary of unknown typing constructs (best-effort)
    if (unknownTypeNamesCollector.size > 0) {
      const entries = Array.from(unknownTypeNamesCollector.entries()).sort((a, b) => b[1] - a[1]);
      const unkList = entries
        .slice(0, 25)
        .map(([n, c]) => `${n}:${c}`)
        .join(', ');
      warnings.push(
        `Module ${baseName}: unknown typing constructs encountered: ${unkList}${entries.length > 25 ? 'â€¦' : ''}`
      );
      // Write JSON report
      try {
        const reportsDir = pathUtils.join('.tywrap', 'reports');
        await fsUtils.writeFile(
          pathUtils.join(reportsDir, `${baseName}.json`),
          JSON.stringify({
            module: baseName,
            unknowns: Object.fromEntries(entries),
            generatedAt: new Date().toISOString(),
          })
        );
      } catch {
        // ignore
      }
    }
  }

  return { written, warnings };
}

/**
 * Invoke the Python IR CLI to get JSON IR for a module.
 */
async function fetchPythonIr(moduleName: string, pythonPath: string): Promise<unknown | null> {
  if (!processUtils.isAvailable()) {
    return null;
  }
  try {
    const result = await processUtils.exec(pythonPath, [
      '-m',
      'tywrap_ir',
      '--module',
      moduleName,
      '--no-pretty',
    ]);
    if (result.code !== 0) {
      // Fallback to invoking local __main__.py
      const localMain = pathUtils.join(process.cwd(), 'tywrap_ir', 'tywrap_ir', '__main__.py');
      const fallback = await processUtils.exec(pythonPath, [
        localMain,
        '--module',
        moduleName,
        '--no-pretty',
      ]);
      if (fallback.code !== 0) {
        return null;
      }
      return JSON.parse(fallback.stdout);
    }
    return JSON.parse(result.stdout);
  } catch {
    return null;
  }
}

/**
 * Convert JSON IR from Python into the internal TypeScript model used by the generator.
 */
function transformIrToTsModel(ir: unknown): TSPythonModule {
  const obj: Record<string, unknown> =
    typeof ir === 'object' && ir !== null ? (ir as Record<string, unknown>) : {};
  const functions = (obj.functions as unknown[]) ?? [];
  const classes = (obj.classes as unknown[]) ?? [];
  const mapParam = (p: Record<string, unknown>): Parameter => ({
    name: String(p.name ?? ''),
    type: parseAnnotationToPythonType(p.annotation),
    optional: Boolean(p.default),
    varArgs: p.kind === 'VAR_POSITIONAL',
    kwArgs: p.kind === 'VAR_KEYWORD',
  });

  const mapFunc = (f: Record<string, unknown>): PythonFunction => ({
    name: String(f.name ?? ''),
    signature: {
      parameters: Array.isArray(f.parameters)
        ? (f.parameters as unknown[]).map(v => mapParam((v ?? {}) as Record<string, unknown>))
        : [],
      returnType: parseAnnotationToPythonType(f.returns),
      isAsync: Boolean(f.is_async),
      isGenerator: Boolean(f.is_generator),
    },
    docstring: (f.docstring as string | undefined) ?? undefined,
    decorators: [],
    isAsync: Boolean(f.is_async),
    isGenerator: Boolean(f.is_generator),
    returnType: parseAnnotationToPythonType(f.returns),
    parameters: Array.isArray(f.parameters)
      ? (f.parameters as unknown[]).map(v => mapParam((v ?? {}) as Record<string, unknown>))
      : [],
  });

  const mapClass = (c: Record<string, unknown>): PythonClass => ({
    name: String(c.name ?? ''),
    bases: Array.isArray(c.bases) ? (c.bases as string[]) : [],
    methods: Array.isArray(c.methods)
      ? (c.methods as unknown[]).map(v => mapFunc((v ?? {}) as Record<string, unknown>))
      : [],
    properties: Array.isArray(c.fields)
      ? ((c.fields as unknown[]).map(v => {
          const p = (v ?? {}) as Record<string, unknown>;
          const optional = Boolean(p.default);
          return {
            name: String(p.name ?? ''),
            type: parseAnnotationToPythonType(p.annotation),
            readonly: false,
            setter: false,
            getter: true,
            optional,
          } as unknown as never;
        }) as unknown as PythonClass['properties'])
      : [],
    docstring: (c.docstring as string | undefined) ?? undefined,
    decorators: (c.typed_dict as boolean) ? ['__typed_dict__'] : [],
    kind: (c.typed_dict as boolean)
      ? 'typed_dict'
      : (c.is_protocol as boolean)
        ? 'protocol'
        : (c.is_namedtuple as boolean)
          ? 'namedtuple'
          : (c.is_dataclass as boolean)
            ? 'dataclass'
            : (c.is_pydantic as boolean)
              ? 'pydantic'
              : 'class',
  });

  const moduleModel: TSPythonModule = {
    name: (obj.module as string) ?? 'module',
    path: undefined,
    version:
      typeof (obj.metadata as Record<string, unknown> | undefined)?.package_version === 'string'
        ? ((obj.metadata as Record<string, unknown> | undefined)?.package_version as string)
        : undefined,
    functions: functions.map(v => mapFunc((v ?? {}) as Record<string, unknown>)),
    classes: classes.map(v => mapClass((v ?? {}) as Record<string, unknown>)),
    imports: [],
    exports: [],
  };
  return moduleModel;
}

function unknownType(): PythonType {
  return { kind: 'primitive', name: 'None' };
}

// Parse Python IR annotation string -> PythonType
function parseAnnotationToPythonType(annotation: unknown, depth = 0): PythonType {
  if (annotation === null || annotation === undefined) {
    return unknownType();
  }
  if (depth > 100) {
    return unknownType();
  }
  const raw = String(annotation).trim();

  // Handle built-in class repr: <class 'int'>
  const classMatch = raw.match(/^<class ['"][^'"]+['"]>$/);
  if (classMatch) {
    const inner = (raw.match(/^<class ['"]([^'"]+)['"]>$/) ?? [])[1] ?? '';
    const name = (inner.split('.').pop() ?? '').toString();
    return mapSimpleName(name);
  }

  // PEP 604 unions: int | str | None
  if (raw.includes(' | ')) {
    const parts = splitTopLevel(raw, '|');
    const types = parts.map(p => parseAnnotationToPythonType(p.trim(), depth + 1));
    return { kind: 'union', types };
  }

  // typing.Union[...]
  if (raw.startsWith('typing.Union[') || raw.startsWith('Union[')) {
    const inner = raw.slice(raw.indexOf('[') + 1, raw.lastIndexOf(']'));
    const parts = splitTopLevel(inner, ',');
    const types = parts.map(p => parseAnnotationToPythonType(p.trim(), depth + 1));
    return { kind: 'union', types };
  }

  // Optional[T]
  if (raw.startsWith('typing.Optional[') || raw.startsWith('Optional[')) {
    const inner = raw.slice(raw.indexOf('[') + 1, raw.lastIndexOf(']'));
    const base = parseAnnotationToPythonType(inner, depth + 1);
    return { kind: 'optional', type: base };
  }

  // Literal[...] -> literal values union
  if (raw.startsWith('typing.Literal[') || raw.startsWith('Literal[')) {
    const inner = raw.slice(raw.indexOf('[') + 1, raw.lastIndexOf(']'));
    const parts = splitTopLevel(inner, ',');
    if (parts.length === 1) {
      return mapLiteral(String(parts[0] ?? '').trim());
    }
    return { kind: 'union', types: parts.map(p => mapLiteral(String(p).trim())) } as PythonType;
  }

  // typing_extensions wrappers: ClassVar[T], Final[T], Required[T], NotRequired[T]
  const extMatch = raw.match(
    /^(typing\.|typing_extensions\.)?(ClassVar|Final|Required|NotRequired)\[(.*)\]$/
  );
  if (extMatch) {
    const inner = extMatch[3] ?? '';
    const base = parseAnnotationToPythonType(inner, depth + 1);
    return base;
  }

  // LiteralString
  if (raw === 'typing.LiteralString' || raw === 'LiteralString') {
    return { kind: 'primitive', name: 'str' } as PythonType;
  }

  // Callable[[...], R]
  if (raw.startsWith('typing.Callable[') || raw.startsWith('Callable[')) {
    const inner = raw.slice(raw.indexOf('[') + 1, raw.lastIndexOf(']'));
    const parts = splitTopLevel(inner, ',');
    if (parts.length >= 2) {
      const paramsPart = (parts[0] ?? '').trim();
      const returnPart = parts.slice(1).join(',').trim();
      const paramInner =
        paramsPart.startsWith('[') && paramsPart.endsWith(']') ? paramsPart.slice(1, -1) : '';
      const paramTypes = ((): PythonType[] => {
        const trimmed = paramInner.trim();
        if (trimmed === '...' || trimmed === 'Ellipsis') {
          return [{ kind: 'custom', name: '...' } as PythonType];
        }
        return trimmed
          ? splitTopLevel(trimmed, ',').map(p => parseAnnotationToPythonType(p.trim(), depth + 1))
          : [];
      })();
      const returnType = parseAnnotationToPythonType(returnPart, depth + 1);
      return { kind: 'callable', parameters: paramTypes, returnType } as PythonType;
    }
  }

  // Mapping[K, V] / Dict[K, V] normalization
  if (raw.startsWith('typing.Mapping[') || raw.startsWith('Mapping[')) {
    const inner = raw.slice(raw.indexOf('[') + 1, raw.lastIndexOf(']'));
    const parts = splitTopLevel(inner, ',');
    const itemTypes = parts.map(p => parseAnnotationToPythonType(p.trim(), depth + 1));
    return { kind: 'collection', name: 'dict', itemTypes } as PythonType;
  }

  // Annotated[T, ...] -> annotated node with base and metadata
  if (raw.startsWith('typing.Annotated[') || raw.startsWith('Annotated[')) {
    const inner = raw.slice(raw.indexOf('[') + 1, raw.lastIndexOf(']'));
    const parts = splitTopLevel(inner, ',');
    if (parts.length > 0) {
      const base = parseAnnotationToPythonType((parts[0] ?? '').trim(), depth + 1);
      const metaParts = parts.slice(1).map(p => String(p).trim());
      return { kind: 'annotated', base, metadata: metaParts } as PythonType;
    }
  }

  // Collections: list[T], dict[K,V], tuple[...], set[T]
  const coll = normalizeCollectionName(raw);
  if (coll) {
    const { name, inner } = coll;
    const itemParts =
      name === 'dict' ? splitTopLevel(inner ?? '', ',') : splitTopLevel(inner ?? '', ',');
    const itemTypes = (inner ? itemParts : []).map(p =>
      parseAnnotationToPythonType(p.trim(), depth + 1)
    );
    return { kind: 'collection', name, itemTypes };
  }

  // typing.Callable[...] unlikely for now - treat as custom

  // Bare names like int, str, float, bool, bytes, None
  return mapSimpleName(raw);
}

function mapSimpleName(name: string): PythonType {
  const n = name.replace(/^typing\./, '').trim();
  if (n === 'int' || n === 'float' || n === 'str' || n === 'bool' || n === 'bytes') {
    // cast to the specific union type without using any
    return { kind: 'primitive', name: n };
  }
  // Track unknown typing-ish names for diagnostics
  if (
    n === 'Any' ||
    n === 'Never' ||
    n === 'LiteralString' ||
    n === 'ClassVar' ||
    n === 'Final' ||
    n === 'TypeAlias' ||
    n === 'Required' ||
    n === 'NotRequired'
  ) {
    try {
      recordUnknown(n);
    } catch {}
  }
  if (n === 'None' || n.toLowerCase() === 'nonetype') {
    return { kind: 'primitive', name: 'None' };
  }
  if (n === 'list' || n === 'List') {
    return { kind: 'collection', name: 'list', itemTypes: [] };
  }
  if (n === 'dict' || n === 'Dict') {
    return { kind: 'collection', name: 'dict', itemTypes: [] };
  }
  if (n === 'tuple' || n === 'Tuple') {
    return { kind: 'collection', name: 'tuple', itemTypes: [] };
  }
  if (n === 'set' || n === 'Set') {
    return { kind: 'collection', name: 'set', itemTypes: [] };
  }
  return { kind: 'custom', name: n };
}

function normalizeCollectionName(
  raw: string
): { name: 'list' | 'dict' | 'tuple' | 'set' | 'frozenset'; inner?: string } | null {
  const m = raw.match(/^(typing\.)?(List|Dict|Tuple|Set|list|dict|tuple|set)\[(.*)\]$/);
  if (!m) {
    return null;
  }
  const nameRaw = String(m[2] ?? '');
  const name =
    nameRaw.toLowerCase() === 'list'
      ? 'list'
      : nameRaw.toLowerCase() === 'dict'
        ? 'dict'
        : nameRaw.toLowerCase() === 'tuple'
          ? 'tuple'
          : 'set';
  return { name, inner: String(m[3] ?? '') };
}

function splitTopLevel(input: string, sep: '|' | ','): string[] {
  const results: string[] = [];
  let level = 0;
  let cur = '';
  let guard = 0;
  for (let i = 0; i < input.length; i++) {
    // Depth/length guard to avoid pathological recursion/loops
    guard++;
    if (guard > 20000) {
      // Bail out safely; push remaining and break
      if (cur.trim()) {
        results.push(cur.trim());
      }
      break;
    }
    const ch = input.charAt(i);
    if (ch === '[' || ch === '(') {
      level++;
    }
    if (ch === ']' || ch === ')') {
      level = Math.max(0, level - 1);
    }
    if (level === 0 && input.slice(i, i + sep.length) === sep) {
      results.push(cur.trim());
      cur = '';
      i += sep.length - 1;
      continue;
    }
    cur += ch;
  }
  if (cur.trim()) {
    results.push(cur.trim());
  }
  return results;
}

function mapLiteral(text: string): PythonType {
  // strip quotes if present
  const t = text.trim();
  if ((t.startsWith("'") && t.endsWith("'")) || (t.startsWith('"') && t.endsWith('"'))) {
    return { kind: 'literal', value: t.slice(1, -1) } as PythonType;
  }
  if (t === 'True' || t === 'False') {
    return { kind: 'literal', value: t === 'True' } as PythonType;
  }
  if (t === 'None') {
    return { kind: 'literal', value: null } as PythonType;
  }
  const num = Number(t);
  if (!Number.isNaN(num)) {
    return { kind: 'literal', value: num } as PythonType;
  }
  // fallback
  return { kind: 'custom', name: t } as PythonType;
}

/**
 * Compute a stable cache key filename for a module and options
 */
async function computeCacheKey(
  moduleName: string,
  options: Partial<TywrapOptions>
): Promise<string> {
  const modules = options.pythonModules ?? {};
  const foundEntry = Object.entries(modules).find(([name]) => name === moduleName);
  const moduleConfig = foundEntry ? foundEntry[1] : undefined;
  const runtimePython = options.runtime?.node?.pythonPath ?? 'python3';
  const keyObject = {
    module: moduleName,
    moduleVersion: moduleConfig?.version ?? null,
    runtime: { pythonPath: runtimePython },
    output: {
      format: options.output?.format ?? 'esm',
      declaration: options.output?.declaration ?? false,
      sourceMap: options.output?.sourceMap ?? false,
    },
    performance: {
      caching: options.performance?.caching ?? false,
      compression: options.performance?.compression ?? 'none',
    },
    typeHints: moduleConfig?.typeHints ?? 'strict',
  } as const;
  const normalized = JSON.stringify(keyObject);
  const digest = await hashUtils.sha256Hex(normalized);
  return `ir_${moduleName}_${digest.slice(0, 16)}.json`;
}

/**
 * Very lightweight .d.ts emitter derived from generated TS wrappers
 * This is intentionally minimal and stable for our wrappers shape.
 */
function renderDts(generatedTs: string): string {
  const header = `// Generated by tywrap\n// Type Declarations\n// DO NOT EDIT MANUALLY\n\n`;
  const lines: string[] = [header];
  // Extract function exports
  const funcRegex = /export\s+async\s+function\s+(\w+)\s*\(([^)]*)\)\s*:\s*Promise<([^>]+)>/g;
  let m: RegExpExecArray | null;
  while ((m = funcRegex.exec(generatedTs)) !== null) {
    const name = m[1];
    const params = m[2];
    const ret = m[3];
    lines.push(`export function ${name}(${params}): Promise<${ret}>;`);
  }
  // Extract class exports and methods
  const classRegex = /export\s+class\s+(\w+)\s*\{([\s\S]*?)\n\}/g;
  while ((m = classRegex.exec(generatedTs)) !== null) {
    const className = String(m[1] ?? '');
    const body = String(m[2] ?? '');
    // Constructor is always variadic unknown[] in current generator
    const methods: string[] = [];
    const methodRegex = /\n\s+async\s+(\w+)\s*\(([^)]*)\)\s*:\s*Promise<([^>]+)>/g;
    let mm: RegExpExecArray | null;
    while ((mm = methodRegex.exec(body)) !== null) {
      methods.push(`  ${mm[1]}(${mm[2]}): Promise<${mm[3]}>;`);
    }
    lines.push(`export class ${className} {`);
    lines.push(`  constructor(...args: unknown[]);`);
    if (methods.length > 0) {
      lines.push(methods.join('\n'));
    }
    lines.push('}');
  }
  return `${lines.join('\n')}\n`;
}

/**
 * Minimal source map placeholder (stable, empty mappings)
 */
function renderSourceMapPlaceholder(moduleName: string | undefined): string {
  const safe = moduleName ?? 'module';
  const map = {
    version: 3,
    file: `${safe}.generated.ts`,
    sources: [],
    names: [],
    mappings: '',
  } as const;
  return JSON.stringify(map, null, 0);
}
