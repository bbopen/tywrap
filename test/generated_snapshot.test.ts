import { describe, it, expect } from 'vitest';
import { generate } from '../src/tywrap.js';
import { fsUtils } from '../src/utils/runtime.js';

describe('Generated TS snapshot - math', () => {
  it('contains stable header and a known function wrapper', async () => {
    const outDir = './generated';
    const res = await generate({
      pythonModules: { math: { runtime: 'node', typeHints: 'strict' } },
      output: { dir: outDir, format: 'esm', declaration: false, sourceMap: false },
      runtime: { node: { pythonPath: 'python3' } },
      performance: { caching: false, batching: false, compression: 'none' },
      development: { hotReload: false, sourceMap: false, validation: 'none' },
    } as any);
    expect(res.written.some(p => p.endsWith('math.generated.ts'))).toBe(true);
    const content = await fsUtils.readFile(`${outDir}/math.generated.ts`);
    expect(content).toContain('Generated by tywrap');
    expect(content).toContain('export async function sqrt');
    expect(content).toContain("__bridge.call('math.sqrt'");
    // basic typing checks
    expect(content).toMatch(/export async function \w+\(.*: .*\): Promise<.*>/);
  });
});
