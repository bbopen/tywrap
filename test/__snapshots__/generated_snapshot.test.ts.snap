// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generated TS snapshot - math > contains stable header and a known function wrapper 1`] = `
"// Generated by tywrap
// Module: math
// DO NOT EDIT MANUALLY

declare const __bridge: { call<T = unknown>(qualified: string, args: unknown[]): Promise<T>; instantiate<T = unknown>(qualified: string, args: unknown[]): Promise<T>; };

/**
 * Return the arc cosine (measured in radians) of x.
 * 
 * The result is between 0 and pi.
 */
export async function acos(x: unknown): Promise<unknown> {
  return __bridge.call('math.acos', [x]);
}

/**
 * Return the inverse hyperbolic cosine of x.
 */
export async function acosh(x: unknown): Promise<unknown> {
  return __bridge.call('math.acosh', [x]);"
`;
